crates/typst/src/model/heading.rs

```diff
@@ -201,15 +215,41 @@ impl Synthesize for Packed<HeadingElem> {
 impl Show for Packed<HeadingElem> {
     #[typst_macros::time(name = "heading", span = self.span())]
     fn show(&self, engine: &mut Engine, styles: StyleChain) -> SourceResult<Content> {
+        const SPACING_TO_NUMBERING: Em = Em::new(0.3);
+
         let span = self.span();
         let mut realized = self.body().clone();
+
+        let hanging_indent = self.hanging_indent(styles);
+
+        let mut indent = match hanging_indent {
+            Smart::Custom(length) => length.resolve(styles),
+            Smart::Auto => Abs::zero(),
+        };
+
         if let Some(numbering) = (**self).numbering(styles).as_ref() {
-            realized = Counter::of(HeadingElem::elem())
+            let numbering = Counter::of(HeadingElem::elem())
                 .display_at_loc(engine, self.location().unwrap(), styles, numbering)?
-                .spanned(span)
-                + HElem::new(Em::new(0.3).into()).with_weak(true).pack()
+                .spanned(span);
+
+            if hanging_indent.is_auto() {
+                let pod = Regions::one(Axes::splat(Abs::inf()), Axes::splat(false));
+                let size = numbering.measure(engine, styles, pod)?.into_frame().size();
+
+                let min_indent =
+                    Length { abs: size.x, em: SPACING_TO_NUMBERING }.resolve(styles);
+                indent = indent.max(min_indent);
```

Comment:
This code works, because `indent` is always `zero` at this point, but is a bit confusing. I think it could just be `indent = size.x + SPACING_TO_NUMBERING.resolve(styles);`